### PR TITLE
validate launch template related labels on provisioner

### PR DESCRIPTION
**Pull request by JacobGabrielson**
Issue #, if available: 273

Description of changes:

Also,

- remove some dead code
- fix some lint warnings
- label launch templates created with KarpenterTagKeyFormat

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by prateekgogia — APPROVED** [review]
LGTM!

File: pkg/cloudprovider/aws/launchtemplate.go [hunk 1]
```diff
@@ -121,10 +121,16 @@ func (p *LaunchTemplateProvider) createLaunchTemplate(ctx context.Context, clust
...
-					Key:   aws.String(fmt.Sprintf(ClusterTagKeyFormat, cluster.Name)),
-					Value: aws.String("owned"),
-				}},
+				Tags: []*ec2.Tag{
+					{
+						Key:   aws.String(fmt.Sprintf(ClusterTagKeyFormat, cluster.Name)),
+						Value: aws.String("owned"),
+					},
+					{
+						Key:   aws.String(fmt.Sprintf(KarpenterTagKeyFormat, cluster.Name)),
```
Comment (ellistarn) [inline pkg/cloudprovider/aws/launchtemplate.go:130]: Can you explain this change?

File: pkg/cloudprovider/aws/suite_test.go [hunk 2]
```diff
@@ -266,7 +266,19 @@ var _ = Describe("Allocation", func() {
 					Expect(env.Client.Create(context.Background(), provisioner)).ToNot(Succeed())
 				}
 			})
+			It("should not fail for unrestricted labels", func() {
```
Comment (ellistarn) [inline pkg/cloudprovider/aws/suite_test.go:269]: "should recognize well known labels"

File: pkg/cloudprovider/aws/launchtemplate.go (hunk 1 above)
Comment (JacobGabrielson) [reply pkg/cloudprovider/aws/launchtemplate.go:130]: Yep, `aws/factory.go` contains:

```go
	// KarpenterTagKeyFormat is set on all Karpenter owned resources.
	KarpenterTagKeyFormat = "karpenter.sh/cluster/%s"
```

So it seems like we are supposed to set this tag on any AWS resources created by Karpenter?

File: pkg/cloudprovider/aws/suite_test.go (hunk 2 above)
Comment (JacobGabrielson) [reply pkg/cloudprovider/aws/suite_test.go:269]: Ok!

File: pkg/cloudprovider/aws/validation.go [hunk 3]
```diff
@@ -36,10 +38,17 @@ func (c *Capacity) validateLabels() error {
 				if !functional.ContainsString(capacityTypes, value) {
 					return fmt.Errorf("%s must be one of %v", key, capacityTypes)
 				}
+			case launchTemplateIdLabel:
```
Comment (ellistarn) [inline pkg/cloudprovider/aws/validation.go:41]: Not a fan of defining booleans, mutating them, and then checking them after the switch. This logic will quickly get out of hand as we add more complexity. I was wary about using a switch statement for validation for this exact reason. 

Consider 
```
f.ValidateAll(
 func() error { validateCapacityLabel(labels) },
 func() error { validateLaunchTemplateLabel(labels) },
)
```
If no label, simply pass the check.

Comment (JacobGabrielson) [reply pkg/cloudprovider/aws/validation.go:41]: Aye, aye, cap'n!

File: pkg/cloudprovider/fake/factory.go [hunk 4]
```diff
@@ -25,10 +25,6 @@ var (
 	NotImplementedError = fmt.Errorf("provider is not implemented. Are you running the correct release for your cloud provider?")
 )
 
-const (
```
[the remark this answers is not in the document]
Comment (JacobGabrielson) [reply pkg/cloudprovider/fake/factory.go:28]: I'm confused by this one?

**Review by ellistarn — DISMISSED** [review]
image

File: pkg/cloudprovider/aws/validation.go [hunk 5]
```diff
@@ -24,22 +24,41 @@ import (
...
+	return functional.ValidateAll(
+		c.validateAllowedLabels,
+		c.validateCapacityLabel,
+		c.validateLaunchTemplateLabels,
+	)
 }
 
-func (c *Capacity) validateLabels() error {
-	for key, value := range c.spec.Labels {
+func (c *Capacity) validateCapacityLabel() error {
```
Comment (bwagner5) [inline pkg/cloudprovider/aws/validation.go:34]: nit: `validateCapacityTypeLabel()`

**Review by bwagner5 — COMMENTED** [review]
one small nit, but otherwise lgtm!

File: pkg/cloudprovider/aws/validation.go (hunk 5 above)
Comment (JacobGabrielson) [reply pkg/cloudprovider/aws/validation.go]: Fix'd

**Review by bwagner5 — APPROVED** [review]
lgtm! 🥳

**Pull request merged**